pkg/azuredisk/nodeserver_v2_test.go

```diff
@@ -30,13 +31,442 @@ import (
 	"github.com/stretchr/testify/assert"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1alpha1"
 	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
 	"sigs.k8s.io/azuredisk-csi-driver/pkg/azuredisk/mockprovisioner"
 	"sigs.k8s.io/azuredisk-csi-driver/pkg/mounter"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/optimization/mockoptimization"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/util"
 	"sigs.k8s.io/azuredisk-csi-driver/test/utils/testutil"
 )
 
+var (
+	testAzVolumeAttachment = v1alpha1.AzVolumeAttachment{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "test-azvolumeattachment",
+			Namespace: "test-namespace",
+		},
+		Spec: v1alpha1.AzVolumeAttachmentSpec{
+			VolumeID:      "test-volume",
+			NodeName:      fakeNodeID,
+			RequestedRole: v1alpha1.PrimaryRole,
+		},
+	}
+)
+
+func TestNodeStageVolume(t *testing.T) {
+	d, err := newFakeDriverV2(t)
+	d.setPerfOptimizationEnabled(false)
+	assert.NoError(t, err)
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	d.crdProvisioner = mockprovisioner.NewMockCrdProvisioner(ctrl)
+
+	stdVolCap := &csi.VolumeCapability_Mount{
+		Mount: &csi.VolumeCapability_MountVolume{
+			FsType: defaultLinuxFsType,
+		},
+	}
+	volumeContext := map[string]string{
+		consts.FsTypeField: defaultLinuxFsType,
+	}
+	volumeContextWithResize := map[string]string{
+		consts.ResizeRequired: "true",
+	}
+
+	stdVolCapBlock := &csi.VolumeCapability_Block{
+		Block: &csi.VolumeCapability_BlockVolume{},
+	}
+
+	volumeCap := csi.VolumeCapability_AccessMode{Mode: 2}
+	volumeCapWrong := csi.VolumeCapability_AccessMode{Mode: 10}
+	invalidLUN := testAzVolumeAttachment.DeepCopy()
+	invalidLUN.Status = v1alpha1.AzVolumeAttachmentStatus{
+		Detail: &v1alpha1.AzVolumeAttachmentStatusDetail{
+			PublishContext: map[string]string{
+				consts.LUN: "/dev/01",
+			},
+		},
+		State: v1alpha1.Attached,
+	}
+
+	publishContext := testAzVolumeAttachment.DeepCopy()
+	publishContext.Status = v1alpha1.AzVolumeAttachmentStatus{
+		Detail: &v1alpha1.AzVolumeAttachmentStatusDetail{
+			PublishContext: map[string]string{
+				consts.LUN: "/dev/disk/azure/scsi1/lun1",
+			},
+		},
+		State: v1alpha1.Attached,
+	}
+
+	emptyContext := testAzVolumeAttachment.DeepCopy()
+	emptyContext.Status = v1alpha1.AzVolumeAttachmentStatus{
+		Detail: &v1alpha1.AzVolumeAttachmentStatusDetail{
+			PublishContext: map[string]string{},
+		},
+		State: v1alpha1.Attached,
+	}
+
+	blkidAction := func() ([]byte, []byte, error) {
+		return []byte("DEVICE=/dev/sdd\nTYPE=ext4"), []byte{}, nil
+	}
+	fsckAction := func() ([]byte, []byte, error) {
+		return []byte{}, []byte{}, nil
+	}
+	resize2fsAction := func() ([]byte, []byte, error) {
+		return []byte{}, []byte{}, nil
+	}
+
+	tests := []struct {
+		desc               string
+		setupFunc          func(*testing.T, *fakeDriverV2)
```

Comment:
Rather than duplicate NodeStageVolume & NodePublishVolume unit tests into separate v1 & v2 test files since they are largely the same, consider testing for the V2 driver to perform V2-specific setup, e.g.

```go
setup: func(t *testing.T, driver FakeDriver) {
    if driverV2, ok := driver.(*fakeDriverV2); ok {
        // perform V2-specific setup here.
    }
}
```

Reply:
We would not only need  to move `fakeDriverV2` but also definition of `DriverV2` to `azuredisk.go`. Would that be okay?

Reply:
Yes.

Reply:
I realized while working through these changes that moving definition of `fakeDriverV2` and `DriverV2` to non-v2 source files will still not allow `fakeDriverV2` to implement `FakeDriver` interface because it will still be missing all the necessary functions (e.g. those defined under `CSIDriver`) as they remain implemented under `azurediskv2` build tags.

I think leaving the `fakeDriverV2` and `DriverV2` as they are and adding to `FakeDriver` interface some v2 test specific functions (e.g. `getCrdProvisioner`, etc.), and implementing them as empty functions in `fakeDriverV1` would be end up being the simplest fix here. What do you think?

Reply:
Yeah, agreed.